eclair-core/src/test/scala/fr/acinq/eclair/payment/receive/InvoicePurgerSpec.scala

```diff
@@ -0,0 +1,91 @@
+import org.scalatest.funsuite.AnyFunSuiteLike
+
+import scala.concurrent.duration.DurationInt
+
+class InvoicePurgerSpec extends TestKitBaseClass with AnyFunSuiteLike {
+
+  import PaymentsDbSpec._
+
+  test("purge invoices") {
+    val dbs = TestPgDatabases()
+    val db = dbs.db.payments
+    val count = 10
+
+    // create expired invoices
+    val expiredInvoices = Seq.fill(count)(Bolt11Invoice(Block.TestnetGenesisBlock.hash, Some(100 msat), randomBytes32(), alicePriv, Left("expired invoice"), CltvExpiryDelta(18),
+      timestamp = 1 unixsec))
+    val expiredPayments = expiredInvoices.map( invoice => IncomingPayment(invoice, randomBytes32(), PaymentType.Standard, invoice.createdAt.toTimestampMilli, IncomingPaymentStatus.Expired))
+    expiredInvoices.lazyZip(expiredPayments).foreach((invoice, payment) => db.addIncomingPayment(invoice, payment.paymentPreimage))
+
+    // create pending invoices
+    val pendingInvoices = Seq.fill(count)(Bolt11Invoice(Block.TestnetGenesisBlock.hash, Some(100 msat), randomBytes32(), alicePriv, Left("pending invoice"), CltvExpiryDelta(18),
+      timestamp = TimestampSecond.now() - 600))
+    val pendingPayments = pendingInvoices.map( invoice => IncomingPayment(invoice, randomBytes32(), PaymentType.Standard, invoice.createdAt.toTimestampMilli, IncomingPaymentStatus.Pending))
+    pendingInvoices.lazyZip(pendingPayments).foreach((invoice, payment) => db.addIncomingPayment(invoice, payment.paymentPreimage))
+
+    // create paid invoices
+    val receivedAt = TimestampMilli.now() + 1.milli
+    val paidInvoices = Seq.fill(count)(Bolt11Invoice(Block.TestnetGenesisBlock.hash, Some(100 msat), randomBytes32(), alicePriv, Left("paid invoice"), CltvExpiryDelta(18),
+      timestamp = TimestampSecond.now()))
+    val paidPayments = paidInvoices.map( invoice => IncomingPayment(invoice, randomBytes32(), PaymentType.Standard, invoice.createdAt.toTimestampMilli, IncomingPaymentStatus.Received(100 msat, receivedAt)))
+    paidInvoices.lazyZip(paidPayments).foreach((invoice, payment) => { db.addIncomingPayment(invoice, payment.paymentPreimage)
+      // receive payment
+      db.receiveIncomingPayment(invoice.paymentHash, 100 msat, receivedAt)
+    })
+
+    val now = TimestampMilli.now()
+    assert(db.listIncomingPayments(0 unixms, now) === expiredPayments ++ pendingPayments ++ paidPayments)
+    assert(db.listIncomingPayments(expiredInvoices.last.createdAt.toTimestampMilli, now) === pendingPayments ++ paidPayments)
+    assert(db.listPendingIncomingPayments(0 unixms, now) === pendingPayments)
+    assert(db.listReceivedIncomingPayments(0 unixms, now) === paidPayments)
+    assert(db.listExpiredIncomingPayments(0 unixms, now) === expiredPayments)
+
+    val interval = 1 seconds
+
+    val _ = system.spawn(InvoicePurger(db, interval), name = "purge-expired-invoices")
+
+    val probe = TestProbe()
+    system.eventStream.subscribe(probe.ref, classOf[PurgeEvent])
```

Comment:
You should subscribe *before* spawning the actor, otherwise there is a race condition if the actor emits the event before you've subscribed to it (this kind of race conditions can happen on our slow CI machines since we aggressively parallelize our tests)

Reply:
fixed in commit [aedcdf](https://github.com/ACINQ/eclair/pull/2174/commits/aedcdfd38978741c0248575a8df3af4a5ba2f76b)

Reply:
fixed in commit [aedcdf](https://github.com/remyers/eclair/commit/aedcdfd38978741c0248575a8df3af4a5ba2f76b)